runtime/src/main/java/uk/ac/manchester/tornado/runtime/graph/TornadoExecutionContext.java

```diff
@@ -204,36 +211,41 @@ public void mapAllTo(TornadoDevice mapping) {
         }
     }
 
-    public void addDevice(TornadoAcceleratorDevice device) {
-        devices.add(device);
+    private void checkDeviceListSize(int deviceIndex) {
+        if (deviceIndex >= devices.size()) {
+            for (int i = devices.size(); i <= deviceIndex; i++) {
+                devices.add(null);
+            }
+        }
     }
 
     public void setDevice(int index, TornadoAcceleratorDevice device) {
+        checkDeviceListSize(index);
         devices.set(index, device);
     }
 
     private void assignTask(int index, SchedulableTask task) {
-        if (taskToDevice[index] != -1) {
-            return;
-        }
 
         String id = task.getId();
         TornadoDevice target = task.getDevice();
-        TornadoAcceleratorDevice accelerator = null;
+        TornadoAcceleratorDevice accelerator;
 
         if (target instanceof TornadoAcceleratorDevice) {
             accelerator = (TornadoAcceleratorDevice) target;
         } else {
             throw new TornadoRuntimeException("Device " + target.getClass() + " not supported yet");
         }
 
+        int deviceIndex = devices.indexOf(target);
         info("assigning %s to %s", id, target.getDeviceName());
 
-        int deviceIndex = devices.indexOf(target);
         if (deviceIndex == -1) {
-            deviceIndex = devices.size();
-            devices.add(accelerator);
+            // deviceIndex = devices.size();
+            deviceIndex = task.meta().getDeviceIndex();
+            checkDeviceListSize(deviceIndex);
+            devices.set(deviceIndex, accelerator);
```

Comment:
you can call the `setDevice()` method here.

Reply:
`devices` is a linkedList.

Reply:
I said that because the `setDevice()` method does this functionality:
```
public void setDevice(int index, TornadoAcceleratorDevice device) {
        checkDeviceListSize(index);
        devices.set(index, device);
}
```

Reply:
Ok, I got what you meant. Yes, you are right.